StateView & Flash Accounting guides added (#848)
Uniswap/docs: docs/contracts/v4/guides/12-state-view.mdx

---
title: StateView
---

> # Introduction to StateView

Reviewer: I think we can align the title to use only "Introduction" like the rest of the guides

```suggestion
# Introduction
```

> When building on **Uniswap v4**, you will often need to read pool state for both onchain and offchain use cases. Onchain contracts can directly invoke the **StateLibrary** to execute these reads during transactions, but offchain systems—such as frontends or analytics services—require a deployed contract with view functions. This is where **StateView** comes in.

Reviewer: We could probably link to the contract code when we first mentioned about it?

```suggestion
When building on **Uniswap v4**, you will often need to read pool state for both onchain and offchain use cases. Onchain contracts can directly invoke the [**StateLibrary**](https://github.com/Uniswap/v4-core/blob/main/src/libraries/StateLibrary.sol) to execute these reads during transactions, but offchain systems—such as frontends or analytics services—require a deployed contract with view functions. This is where [**StateView**](https://github.com/Uniswap/v4-periphery/blob/main/src/lens/StateView.sol) comes in.
```

```tsx
// Frontend or analytics client using StateView
const stateView = getContract({
  address: STATE_VIEW_ADDRESS,
  abi: stateViewABI
});

const { sqrtPriceX96 } = await stateView.read.getSlot0([poolId]);
// ... use the price in your application ...
```

This separation ensures that each context (onchain vs. offchain) uses the most efficient data reading pattern.

# Usage With Frontend Clients

Frontend applications frequently display real-time information about pools, positions, and other market data—without incurring transaction costs. **StateView** addresses these requirements by exposing read-only functions tailored for offchain integrations.

## Setting Up With Viem

We’ll use [**viem**](https://viem.sh/), a TypeScript library for Ethereum, to demonstrate how to connect to **StateView**.

```tsx
import { createPublicClient, http } from 'viem';
import { mainnet } from 'viem/chains';

// Initialize the client
const client = createPublicClient({
  chain: mainnet,
  transport: http()
});

// Set up StateView contract instance
const stateView = getContract({
  address: STATE_VIEW_ADDRESS,
  abi: stateViewABI,
  client
})
```

> > **Note:** The stateView object comes from our getContract call above. Make sure you’ve imported stateViewABI correctly before attempting to read from the contract.
> > 

Reviewer: Could we make this italic like the above?

```suggestion
> **Note:** _The stateView object comes from our getContract call above. Make sure you’ve imported stateViewABI correctly before attempting to read from the contract._
> 
```

With this setup, you can now:

- **Connect to an Ethereum network**
- **Call StateView’s read functions**
- **Retrieve pool information offchain at no gas cost**

### Handling Errors and Invalid Pool IDs

When calling `stateView.read.<function>([poolId])`, be mindful that:

> - If you pass an invalid `poolId` (usually a `bytes32` in Uniswap v4), the call may revert or return unexpected data.

Reviewer: "typically" seems better and aligned in this doc, also is it better we link to the contract code for additional context on byte32 as well?

```suggestion
- If you pass an invalid `poolId` (typically a [`bytes32`](https://github.com/Uniswap/v4-core/blob/main/src/types/PoolId.sol#L6) in Uniswap v4), the call may revert or return unexpected data.
```

  // getSlot0 returns:
  // - Current price (sqrtPriceX96) in Q64.96 fixed-point format
  // - Active tick
  // - Protocol and LP fee settings
  const {
    sqrtPriceX96,
    tick,
    protocolFee,
    lpFee
  } = await stateView.read.getSlot0([poolId]);

  return {
    price: calculatePrice(sqrtPriceX96), // implement your math logic for Q64.96
    tick,
    protocolFee,
    lpFee
  };
};
```

**What it Returns:**

- **`sqrtPriceX96`**: The current pool price in Q64.96 fixed-point format.
- **`tick`**: The current tick in which the pool is operating.
- **`protocolFee`** and **`lpFee`**: Fee parameters for protocol and LP fee tiers.

## Getting Pool Liquidity

To understand how much liquidity a pool holds:

```tsx
// Example: Reading the total active liquidity of a pool
const getPoolLiquidity = async (poolId: string) => {
  // getLiquidity returns the total liquidity currently active in the pool
  const liquidity = await stateView.read.getLiquidity([poolId]);
  return liquidity;
};
```

**Why It Matters:**

- Helps gauge the depth of the pool
- Influences price impact calculations in trading
- Provides context for the pool’s capacity to absorb trades

> 

Reviewer: Should we just remove this example? Since it's a relatively simple one and we already provided one example above and explained the `getLiquidity` function in the next section as well.

Reviewer 2: fine to keep, while simple i think its a useful callout to the function itself

# Core Functions and Return Types

While **StateView** exposes many functions, here are several essential calls for most offchain applications. Each function typically takes a `poolId` (of type `bytes32`) as the key input, identifying which pool to query.

> 1. **`getSlot0(poolId)`**

Reviewer 3: Could you link these functions to the technical reference documentation? https://docs.uniswap.org/contracts/v4/reference/periphery/lens/StateView#getslot0

Same for the subsequent functions.

I like how you mention the possibilities of what you can do with this function

Reviewer 2: agreed, we should link out to `/contracts/v4/reference/periphery/lens/StateView#getslot0` so developers have a better understanding of the parameters

Author: done

    - Returns `(uint160 sqrtPriceX96, int24 tick, uint8 protocolFee, uint8 lpFee)`.
    - Essential for displaying real-time price data and fees.
2. **`getLiquidity(poolId)`**
    - Returns `uint128 liquidity` (the total active pool liquidity).
    - Used to assess trading depth and volatility.
3. **`getPositionInfo(positionId)`**
    - Returns `(uint128 liquidity, uint256 feeGrowthInside0Last, uint256 feeGrowthInside1Last)`.
    - Critical for tracking user positions, especially to calculate earned fees over time.
4. **`getFeeGrowthGlobals(poolId)`**
    - Returns `(uint256 feeGrowthGlobal0, uint256 feeGrowthGlobal1)`.
    - Useful for analytics around total fee accumulation in the pool.

### Note on `poolId` and `positionId`

> - In **Uniswap v4**, a `poolId` is typically a `bytes32` that can be derived from the pair of tokens, fee tier, and other parameters.

Reviewer 3: Could be more specific. 
`keccak256(abi.encode(poolKey))` where the poolKey consists of the following fields - https://github.com/Uniswap/v4-core/blob/19833c4951439bbf6ef09dee573a62127601440b/src/types/PoolKey.sol#L4

Author: done
